File: src/python/sorting/counting_sort.py
def counting_sort(arr):
    """
    Performs counting sort on a list of non-negative integers.

    Args:
        arr (list): The list of non-negative integers to sort.

    Returns:
        list: The sorted list.
    """
    if not arr:
        return []

    max_val = max(arr)
    count = [0] * (max_val + 1)
    output = [0] * len(arr)

    # Store count of each character
    for num in arr:
        count[num] += 1

    # Store cumulative count
    for i in range(1, max_val + 1):
        count[i] += count[i - 1]

    # Place elements in output array
    for num in reversed(arr):
        output[count[num] - 1] = num
        count[num] -= 1

    return output

File: c/src/counting_sort.c
#include "../sorting_algorithms.h"
#include <stdlib.h>
#include <string.h>

void counting_sort(int *arr, size_t n) {
    if (n == 0) return;
    int min = arr[0], max = arr[0];
    for (size_t i = 1; i < n; ++i) {
        if (arr[i] < min) min = arr[i];
        if (arr[i] > max) max = arr[i];
    }
    size_t range = (size_t)(max - min + 1);
    int *count = (int *)calloc(range, sizeof(int));
    int *output = (int *)malloc(n * sizeof(int));
    if (!count || !output) {
        free(count);
        free(output);
        return;
    }
    for (size_t i = 0; i < n; ++i) count[arr[i] - min]++;
    for (size_t i = 1; i < range; ++i) count[i] += count[i - 1];
    for (size_t i = n; i-- > 0;) output[--count[arr[i] - min]] = arr[i];
    memcpy(arr, output, n * sizeof(int));
    free(count);
    free(output);
}
